stockroom/diff.py

```diff
@@ -0,0 +1,126 @@
+
+        # Generate rows
+        for row in self.all_keys:
+            ref_row = ""
+            dest_row = ""
+
+            if row in self._added:
+                dest_row = "[green] %s [/green]"%(self._added[row])
+
+            elif row in self._mutated:
+                ref_row = "[red] %s [/red]"%(self._mutated[row][0])
+                dest_row = "[red] %s [/red]"%(self._mutated[row][1])
+
+            elif row in self._deleted:
+                ref_row = "[green] %s [/green]"%(self._deleted[row])
+
+            table.add_row(row, ref_row, dest_row)
+        return table
+
+    def parse_diff(self):
+        """
+        Parse the hangar.diff and store the changes in the TagDiff object
+        """
+
+        print('Accessing ref_commit')
+        ref_checkout = self.repo.checkout(commit = self.ref_commit)
+
+        print("Accessing def_commit")
+        dest_checkout = self.repo.checkout(commit = self.dest_commit)
+        diff = ref_checkout.diff.commit(self.dest_commit).diff
+
+        # Get all the keys in both checkouts
+        all_keys = set()
+
+        for tag in ref_checkout.metadata.keys():
+            key = tag.split(SEP)[-1]
+            if key != "type":
+                all_keys.add(key)
+
+        for tag in dest_checkout.metadata.keys():
+            key = tag.split(SEP)[-1]
+            if key not in all_keys and key != "type":
+                all_keys.add(key)
+        self.all_keys = all_keys
+
+        if len(diff.added.metadata) > 0:
+            for keyRecord  in diff.added.metadata:
+                key = keyRecord.key.split(SEP)[-1]
+                if key != "type":
+                    keyvalue = dest_checkout.metadata[tagkey(key)]
+                    self._added[key] = keyvalue
+
+        if len(diff.mutated.metadata) > 0:
+            for keyRecord in diff.mutated.metadata:
+                key = keyRecord.key.split(SEP)[-1]
+                if key != "type":
+                    keyvalue_ref = ref_checkout.metadata[tagkey(key)]
+                    keyvalue_dest = dest_checkout.metadata[tagkey(key)]
+                    self._mutated[key] = (keyvalue_ref, keyvalue_dest)
+
+        if len(diff.deleted.metadata) > 0:
+            for keyRecord in diff.deleted.metadata:
+                key = keyRecord.key.split(SEP)[-1]
+                if key != "type":
+                    keyvalue_ref = ref_checkout.metadata[tagkey(key)]
+                    self._deleted[key] = keyvalue_ref
+
+    def __repr__(self):
+        output = list()
+        append = output.append
+        console = Console()
+        for line in Segment.split_and_crop_lines(console.render(self.gen_tables(), None), 90):
+            for text, style in line:
+                if style:
+                    append(style.render(text))
+                else:
+                    append(text)
+        return ''.join(output)
+
+    def _repr_html_(self):
+        return print(self.gen_tables())._repr_html_()
```

Comment:
Why do you have to print this here?

Reply:
The print function is used to generate the HTML in Rich